<p align="center";">
  <img src="figures/img/DNB-BINF-Core-logo.png" alt="DNB-BINF-Core logo" width="560px" />
</p>
<p align="center";">

# Single cell ATAC seq BED builder

*This repo is currently under constructionâ€”stay tuned for updates!* ðŸš§

This repo is a simple tool designed to generate BED files that can be added to a 10x scATAC seq reference genome. The three files generated by this pipeline are `blacklist.bed`, `promoters.bed`, and `enhancers.bed`. Including these three files with a 10x scATAC seq reference genome allows `blacklist_region_fragments`, `promoter_region_fragments`, and `enhancer_region_fragments` counts to be calculate and reported in the `singlecell.csv` output file generated when running `cellranger atac` to align 10x single cell atac seq data.

While 10x pre-built reference genomes used to include these BED files by default, they are no longer included in new reference genome releases and are not available for the pre-built `2020` and `2024` 10x scATAC reference genomes as of the [Cell Ranger ATAC 2.0 release](https://www.10xgenomics.com/support/software/cell-ranger-atac/latest/release-notes/release-notes#cell-ranger-atac-2-0-may-3-2021-4017ff).


### Key Features
This tool will allow users to generate properly formatted BED files for any pre-built or custom reference genome by providing an input BED file of blacklist regions (or excluded regions) for the reference genome of interest and a GFF file of regulatory regions for the reference genome release of interest.

### Why Are These BED Files Important?
BED files for promoters, enhancers, and blacklist regions play a key role in scATAC-seq analysis:

**Functional Annotation of Peaks:**

Promoter and enhancer BED files allow annotation of ATAC-seq peaks, helping identify which accessible regions correspond to gene regulatory elements.

**Quality Control Metrics:**

Overlap with promoters and enhancers is used to assess data quality and biological relevance.
Blacklist regions help filter out artefactual signals and improve overall data quality.

### Below is the main directory structure listing the analyses and data files used in this repository

```
â”œâ”€â”€ figures
â”œâ”€â”€ LICENSE
â”œâ”€â”€ README.md
â””â”€â”€ SECURITY.md
```

## Authors

Sharon Freshour, PhD ([@sharonfreshour](https://github.com/sharonfreshour))


## Contact

Contributions, issues, and feature requests are welcome! Please feel free to check [issues](https://github.com/stjude-dnb-binfcore/devops-containers/issues).

---

*These tools and pipelines have been developed by the Bioinformatic core team at the [St. Jude Children's Research Hospital](https://www.stjude.org/). These are open access materials distributed under the terms of the [BSD 2-Clause License](https://opensource.org/license/bsd-2-clause), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.*